x-pack/plugins/fleet/server/services/security/uninstall_token_service/index.ts

```diff
@@ -0,0 +1,412 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { randomBytes, createHash } from 'crypto';
+
+import { chunk } from 'lodash';
+
+import type {
+  SavedObjectsClientContract,
+  SavedObjectsCreatePointInTimeFinderOptions,
+  SavedObjectsBulkUpdateObject,
+  SavedObjectsFindResult,
+} from '@kbn/core-saved-objects-api-server';
+import type {
+  AggregationsMultiBucketAggregateBase,
+  AggregationsTopHitsAggregate,
+} from '@elastic/elasticsearch/lib/api/types';
+import type { EncryptedSavedObjectsClient } from '@kbn/encrypted-saved-objects-plugin/server';
+import type { KibanaRequest } from '@kbn/core-http-server';
+import { SECURITY_EXTENSION_ID } from '@kbn/core-saved-objects-server';
+import { asyncForEach } from '@kbn/std';
+
+import { UNINSTALL_TOKENS_SAVED_OBJECT_TYPE, SO_SEARCH_LIMIT } from '../../../constants';
+import { appContextService } from '../../app_context';
+import { agentPolicyService } from '../../agent_policy';
+
+interface UninstallTokenSOAttributes {
+  policy_id: string;
+  token: string;
+  token_plain: string;
+}
+
+interface UninstallTokenSOAggregationBucket {
+  key: string;
+  latest: AggregationsTopHitsAggregate;
+}
+
+interface UninstallTokenSOAggregation {
+  by_policy_id: AggregationsMultiBucketAggregateBase<UninstallTokenSOAggregationBucket>;
+}
+
+export interface UninstallTokenServiceInterface {
+  getTokenForPolicyId(policyId: string): Promise<string>;
+  getTokensForPolicyIds(policyIds: string[]): Promise<Record<string, string>>;
+  getAllTokens(): Promise<Record<string, string>>;
+  getHashedTokenForPolicyId(policyId: string): Promise<string>;
+  getHashedTokensForPolicyIds(policyIds?: string[]): Promise<Record<string, string>>;
+  getAllHashedTokens(): Promise<Record<string, string>>;
+  generateTokenForPolicyId(policyId: string, force?: boolean): Promise<string>;
+  generateTokensForPolicyIds(policyIds: string[], force?: boolean): Promise<Record<string, string>>;
+  generateTokensForAllPolicies(force?: boolean): Promise<Record<string, string>>;
+  encryptTokens(): Promise<void>;
+}
+
+export class UninstallTokenService implements UninstallTokenServiceInterface {
+  private _soClient: SavedObjectsClientContract | undefined;
+
+  constructor(private esoClient: EncryptedSavedObjectsClient) {}
+
+  /**
+   * gets uninstall token for given policy id
+   *
+   * @param policyId agent policy id
+   * @returns token
+   */
+  public async getTokenForPolicyId(policyId: string): Promise<string> {
+    return (await this.getTokensForPolicyIds([policyId]))[policyId];
+  }
+
+  /**
+   * gets uninstall tokens for given policy ids
+   *
+   * @param policyIds agent policy ids
+   * @returns Record<policyId, token>
+   */
+  public async getTokensForPolicyIds(policyIds: string[]): Promise<Record<string, string>> {
```

Comment:
This function does a lot of things, could you add comments to explain what is happening on each step? Why are we using aggs for getting tokens? I would expect here a simple search by policyIds, but I'm sure I'm missing something.

Reply:
sure. will add comments. aggs are necessary since there are multiple tokens per policy (we're keeping a full history). so we need to make sure we're only getting the newest token for each policy.